### PR TITLE
Implement #812 (brew configure-shell)

### PR DESCRIPTION
**Pull request by jamescostian**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Linuxbrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Linuxbrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

### CODE REVIEW

File: Library/Homebrew/cmd/configure-shell.rb
Comment (maxim-belkin) [inline Library/Homebrew/cmd/configure-shell.rb]: You're coding this in a Bash-specific way... Look at the `export_value` function in `Library/Homebrew/utils/shell.rb` for an inspiration how to code this in a general way for any shell.
Another question: is the goal to make SHELL (Bash, Zsh, whatever) read this file, or Homebrew/brew?

Comment (maxim-belkin) [reply Library/Homebrew/cmd/configure-shell.rb]: I mean, I understand that currently you're prepping it to make SHELL read this, but... should it be this way?

Comment (jamescostian) [reply Library/Homebrew/cmd/configure-shell.rb]: Thanks for that pointer! I'll use `export_value` instead.

And yes, the goal is for SHELL to read this file. The original purpose of this file was simply to set `PATH`, `MANPATH`, and `INFOPATH` - brew doesn't need to read this, SHELL needs to read this.

Comment (sjackman) [reply Library/Homebrew/cmd/configure-shell.rb]: Just as `/etc/environment` is always in `bash` syntax, I'm thinking that `$HOMEBREW_PREFIX/etc/environment` should similarly always be in `bash` syntax. If the user is using `csh` or `fish`, I feel the configuration should be written to `$HOMEBREW_PREFIX/etc/environment.csh` or `$HOMEBREW_PREFIX/etc/environment.fish`. This is necessary if the users uses more than one shell, say `bash` and `fish`. Note `zsh` is compatible with `bash` syntax.

**Comment by sjackman** [comment]
The scope of this PR has grown a bit beyond the original proposal. I'd prefer not to ask or set `HOMEBREW_DEVELOPER` and `HOMEBREW_VERBOSE` in this PR.

File: Library/Homebrew/cmd/configure-shell.rb
Comment (sjackman) [inline Library/Homebrew/cmd/configure-shell.rb]: I suggest…
```ruby
environment = <<~EOS
  # Homebrew environment
  # Created by running brew configure-shell
  export PATH="#{HOMEBREW_PREFIX}/bin:$PATH"
  export MANPATH="#{HOMEBREW_PREFIX}/share/man:$MANPATH"
  export INFOPATH="#{HOMEBREW_PREFIX}/share/info:$INFOPATH"
EOS
File.write HOMEBREW_PREFIX/"etc/environment", environment, perm: 0755
```

File: Library/Homebrew/cmd/configure-shell.rb
Comment (sjackman) [inline Library/Homebrew/cmd/configure-shell.rb]: Needs an end-of-line character at the end as well.
```ruby
      File.write "#{ENV["HOME"]}/.profile", current_profile + "source #{HOMEBREW_PREFIX}/etc/environment # Added by brew configure-shell\n"
```

**Review by sjackman — CHANGES_REQUESTED** [review]
Thanks for this work, James!

**Comment by sjackman** [comment]
> if MANPATH was empty before running this code, then MANPATH will become $HOMEBREW_PREFIX/share/man: - is that trailing comma ok?

I believe when `MANPATH` ends in a colon, `man` also searches the default host directories, which is desirable.

File: Library/Homebrew/cmd/configure-shell.rb
Comment (jamescostian) [reply Library/Homebrew/cmd/configure-shell.rb]: While this solution does look much better, it doesn't account for shell quoting and it won't work with other shells. I think Max's idea to use `export_value` may be better:

```
environment = <<~EOS
  # Homebrew environment
  # Created by running brew configure-shell
  #{export_value "PATH", "#{HOMEBREW_PREFIX}/bin:$PATH", "bash"}
  #{export_value "MANPATH", "#{HOMEBREW_PREFIX}/bin:$MANPATH", "bash"}
  #{export_value "INFOPATH", "#{HOMEBREW_PREFIX}/bin:$INFOPATH", "bash"}
EOS
File.write HOMEBREW_PREFIX/"etc/environment", environment, perm: 0755
```

Then wrap that in a for loop (not sure how to do that in Ruby yet, but I'll google it) and you'll end up writing environment for each of the shells.

Comment (jamescostian) [reply Library/Homebrew/cmd/configure-shell.rb]: On the topic of the chmod permissions, correct me if I'm wrong, but say you're logged in as a user named "Alice" and you write this file, and then "Bob" sources this file (same machine, just a different user). Now "Alice" can change the file manually and do bad things when "Bob" has his `~/.profile` sourced

In order to avoid this, we'd want to `chmod 0755` and `chown root:root` so that only the super user could cause problems for all the users on the system.

Am I just being paranoid or is this something we should prevent?

Comment (sjackman) [reply Library/Homebrew/cmd/configure-shell.rb]: The double quotes above handle shell quoting, assuming that the user does not have a dollar sign or double quote in their path.

Comment (sjackman) [reply Library/Homebrew/cmd/configure-shell.rb]: Anyone sourcing `$HOMEBREW_PREFIX/etc/environment` needs to trust the owner of that file. I'm okay with that. If that person is the sysadmin, then you have to trust the sysadmin, which you have to anyway.

File: Library/Homebrew/cmd/configure-shell.rb
Comment (sjackman) [inline Library/Homebrew/cmd/configure-shell.rb]: I forgot that a shell snippet doesn't need to have execute permissions to be able to `source` it. Please remove `, perm: 0755`.

File: Library/Homebrew/cmd/configure-shell.rb
Comment (sjackman) [inline Library/Homebrew/cmd/configure-shell.rb]: Please change
```diff
-Have the current user's ~/.profile source the environment file
+Modify the current user's ~/.profile to source the environment file.
```

File: Library/Homebrew/utils.rb
Comment (sjackman) [inline Library/Homebrew/utils.rb]: Please revert this chunk.

**Review by sjackman — CHANGES_REQUESTED** [review]

**Comment by sjackman** [comment]
Looking good, James! Please loop over `~/.profile`, `~/.bash_profile`, and `~/.zprofile` and modify each file if it exists. Do not create the file if it does not exist.

**Comment by maxim-belkin** [comment]
how do we want to handle C-shell?

**Comment by sjackman** [comment]
```
brew configure-shell --help
Warning: No help text in: /home/sjackman/.linuxbrew/Homebrew/Library/Homebrew/cmd/configure-shell.rb
```
Please add help text. You can look at other commands in `cmd/*.rb` to see the format.

**Comment by sjackman** [comment]
Please squash all five commits of this PR down to a single commit. You can do that using
```sh
git rebase -i origin/master
git push --force https://github.com/jamescostian/brew
```
When making further changes to this PR, you can use `git commit --amend` followed by `git push --force` to modify the existing commit, rather than creating new ones.

**Comment by sjackman** [comment]
I tested out your current PR, and it works well! This feature will be a nice improvement to the user installation experience. Thanks!

**Comment by jamescostian** [comment]
Alright everything should be taken care of except:

- Currently there is no `$HOMEBREW_PREFIX/etc/environment.csh` or `$HOMEBREW_PREFIX/etc/environment.fish`
- If you run this command, then you comment out `source ...` and then run this command again, your `~/.profile` will still contain your commented out line (but it will also have another, uncommented `source ...` line)
- If you run this command, move the brew installation, and run this command again, your `~/.profile` will still contain the old `source ...` line (but it will also have the new `source ...` line with the correct path)
 
@sjackman would you like me to tackle these in this PR, or another PR, or leave them unfixed?

**Comment by jamescostian** [comment]
Hmmm it has now failed the build twice now saying:

```
Writing markdown to /home/linuxbrew/.linuxbrew/Homebrew/docs/Manpage.md
Writing man page to /home/linuxbrew/.linuxbrew/Homebrew/manpages/brew.1
Writing man page to /home/linuxbrew/.linuxbrew/Homebrew/manpages/brew-cask.1
==> FAILED
```

Does anyone know how to fix that?

**Comment by jonchang** [comment]
`brew man` then commit the changes.

**Comment by sjackman** [comment]
Those three issues that you've raised can remain unaddressed by this PR.

**Review by sjackman — APPROVED** [review]

**Comment by sjackman** [comment]
Looks good to me! Thanks, James!

One thought though: if `$HOMEBREW_PREFIX/etc/environment` already exists, and its contents are not equal to those that are about to be written, it should instead print an error message:

> Error: ~/.linuxbrew/etc/environment already exists. If you wish to replace this file, remove this file and run `brew configure-shell` again.

If the contents are identical, there's no need to print the error message.

Print the error message using `odie "message goes here"` or
```ruby
odie <<~EOS
  message goes here
EOS
```

**Comment by maxim-belkin** [comment]
why do you not use `export_value` and `profile` that are defined in `Library/Homebrew/utils/shell.rb`?

**Comment by jamescostian** [comment]
@jonchang Thanks! I ran `brew man` and took the changes it suggested relating to this command. It also suggested some changes for other commands - should I be accepting those changes as well?

@maxim-belkin I agree, and I suggested that [here](https://github.com/Linuxbrew/brew/pull/829#discussion_r213143859) but I didn't get a green light in that thread - @sjackman what do you think? As Max mentioned, using `export_value` would mean less code duplication and would make it easier to have this command also make `environment.csh` and `environment.fish`

*EDIT*: you could also use `SHELL_PROFILE_MAP` to get a list of files that should source `environment` (if they exist), which would be better for having less code duplication

**Comment by jamescostian** [comment]
@sjackman your request that `environment` should not be overwritten if it exists has been taken care of, and the error message (that only shows if there are different contents) has also been added :tada:

File: Library/Homebrew/cmd/configure-shell.rb
Comment (sjackman) [inline Library/Homebrew/cmd/configure-shell.rb]: `odie` prefixes the message with `Error: ` so you can remove `Error: `

Comment (sjackman) [reply Library/Homebrew/cmd/configure-shell.rb]: Because the message is long, it can be easy to miss seeing the `if` at the end of the line. I suggest…
```ruby
if File.read(env_path) != environment
…
end
```
I also prefer the more typical `File.read(env_path)` to `(File.read env_path)`

**Comment by sjackman** [comment]
Looks good!

Please print informational messages about which shell profile files have been modified. For example…
```
Modified ~/.profile
Skipped ~/.bash_profile because it's already configured
```

Please add to the end…
```ruby
oh1 "To configure your current shell session, please run..."
puts "  source #{HOMEBREW_PREFIX}/etc/environment"
```
This message should be printed whether or not that file is updated.

**Comment by maxim-belkin** [comment]
`linux-pam` installs HOMEBREW_PREFIX/etc/environment

**Comment by sjackman** [comment]
Argh! I was worried that some other package would install that file! Thanks for catching this issue, Maxim. The default contents of the file are empty (except for comments):
```sh
#
# This file is parsed by pam_env module
#
# Syntax: simple "KEY=VAL" pairs on separate lines
#
```
All the same though, I think it'd be safer to use a different file name. I suggest
`#{HOMEBREW_PREFIX}/etc/brew.env`

**Comment by jamescostian** [comment]
Alright, all of that has been taken care of now 👍

**Comment by sjackman** [comment]
Looking good! Two final fixes:

Add `$HOMEBREW_PREFIX/sbin` to the `PATH` after `bin`.

If no shell profile files currently exist, create a new file `~/.profile` and modify it.

**Comment by jamescostian** [comment]
Good points! They are now taken care of

File: Library/Homebrew/cmd/configure-shell.rb
Comment (sjackman) [inline Library/Homebrew/cmd/configure-shell.rb]:
```ruby
if profile_and_friends.empty?
```

File: Library/Homebrew/cmd/configure-shell.rb
Comment (sjackman) [inline Library/Homebrew/cmd/configure-shell.rb]:
```ruby
profile_and_friends = ["#{ENV["HOME"]}/.profile"]
```

**Review by sjackman — CHANGES_REQUESTED** [review]
Two minor style comments and then we're good to merge!

**Comment by sjackman** [comment]
Also note these `brew style` issues:
```
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/test/cmd/configure-shell.rb:19:7: C: Align .and with .to_stdout on line 18.
      .and not_to_output.to_stderr
      ^^^^
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/test/cmd/configure-shell.rb:29:7: C: Align .and with .to_stdout on line 28.
      .and not_to_output.to_stderr
      ^^^^
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/test/cmd/configure-shell.rb:41:7: C: Align .and with .to_stdout on line 40.
      .and not_to_output.to_stderr
      ^^^^
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/test/cmd/configure-shell.rb:45:7: C: Align .and with .to_stdout on line 44.
      .and not_to_output.to_stderr
      ^^^^
```
https://circleci.com/gh/Linuxbrew/brew/1212

**Comment by sjackman** [comment]
Once this feature has made it into a stable release of Linuxbrew/brew, I'll update `Linuxbrew/install` and the `README.md` of `Linuxbrew/brew` to use it!

Once this PR is merged and has some testing with users, you can also submit it upstream to Homebrew/brew, if you like. (I'd appreciate it.) You can mention that it's primarily useful for Linux users, but is also useful for non-/usr/local installs on macOS. I don't have write permissions to `/usr/local` on my machine at work (due to be a clinical laboratory), so I have Homebrew installed in `~/.homebrew`, so this tool is useful on macOS as well in that situation.

File: Library/Homebrew/cmd/configure-shell.rb
Comment (maxim-belkin) [inline Library/Homebrew/cmd/configure-shell.rb:19]: Shall we add Homebrew bash completions here? 
```
source #{HOMEBREW_PREFIX}/Homebrew/completions/bash/brew
```

Comment (maxim-belkin) [reply Library/Homebrew/cmd/configure-shell.rb:19]: Ah, nevermind - this is for `~/.profile`, not `~/.bashrc`...

Comment (sjackman) [reply Library/Homebrew/cmd/configure-shell.rb:19]: Hmm. I didn't know about this feature of Homebrew in fact. Let's merge this PR as is. We could consider completions in a separate PR.
Loading completions can't be done in `/home/linuxbrew/.linuxbrew/etc/brew.env` because that file is loaded by both `bash` and `zsh`. It could be added directly to `~/.bash_profile` and `~/.zprofile`.

Comment (sjackman) [reply Library/Homebrew/cmd/configure-shell.rb:19]: I believe `oh-my-zsh` and `prezto`, which are pretty common for zsh users, already have Homebrew completion enabled.

Comment (sjackman) [reply Library/Homebrew/cmd/configure-shell.rb:19]: > Ah, nevermind - this is for ~/.profile, not ~/.bashrc

Just saw this post after I wrote my response.

Comment (maxim-belkin) [reply Library/Homebrew/cmd/configure-shell.rb:19]: No problem. I forgot this is for `.profile` files -- these are for "login" shells so should not have stuff like completions.

File: Library/Homebrew/cmd/configure-shell.rb
Comment (jamescostian) [reply Library/Homebrew/cmd/configure-shell.rb]: I agree - previously the linter yelled at me for writing `if` statements that weren't on the end of a line, so that's why I made that line so long. But somehow it's not yelling now (I guess I had to put `if` at the end of some *other* line).

File: Library/Homebrew/cmd/configure-shell.rb
Comment (jamescostian) [reply Library/Homebrew/cmd/configure-shell.rb]: AHHHHH that looks much better!

File: Library/Homebrew/cmd/configure-shell.rb
Comment (jamescostian) [reply Library/Homebrew/cmd/configure-shell.rb]: Done

File: Library/Homebrew/cmd/configure-shell.rb
Comment (jamescostian) [reply Library/Homebrew/cmd/configure-shell.rb:19]: Should this conversation be marked as resolved or is there anything that needs to be done to address this?

**Comment by jamescostian** [comment]
@sjackman Thanks! And yeah, after this PR merges I'll make the same one to Homebrew/brew

The `brew style` issues are weird... I don't see them on my machine (I'm doing this development in an Ubuntu docker container), and fixing them requires the code to look pretty ugly, e.g.:

```
    expect { run_cmd }
      .to be_a_success
      .and output(%r{Modified ~\/\.profile}).to_stdout
                                            .and not_to_output.to_stderr
```

Any ideas why this is happening?

File: Library/Homebrew/cmd/configure-shell.rb
Comment (sjackman) [reply Library/Homebrew/cmd/configure-shell.rb:19]: I consider it resolved.

**Comment by sjackman** [comment]
No, I don't know why that's happening. That same style is used in other tests, where it seems fine. To make `brew style` happy, I suggest…
```
    expect { run_cmd }
      .to be_a_success
      .and output(%r{Modified ~\/\.profile})
      .to_stdout
      .and not_to_output.to_stderr
```

**Comment by sjackman** [comment]
```ruby
 .and output(%r{Modified ~\/\.profile})
```
The forward slash `/` does not need to be escaped (I believe).

**Comment by jamescostian** [comment]
Alright those two things should be taken care of now

File: Library/Homebrew/cmd/configure-shell.rb
Comment (maxim-belkin) [inline Library/Homebrew/cmd/configure-shell.rb:47]: Why not
```ruby
File.write profile_path, add_to_profile, mode: 'a'
```
?

File: Library/Homebrew/cmd/configure-shell.rb
Comment (maxim-belkin) [inline Library/Homebrew/cmd/configure-shell.rb:44]: I would suggest checking for lines that match `^\s*source #{env_path}`
```ruby
if File.foreach(profile_path).grep(/^\s*source #{env_path}/).any?
```

File: Library/Homebrew/cmd/configure-shell.rb
Comment (sjackman) [reply Library/Homebrew/cmd/configure-shell.rb:47]: Yep, that's valid too. We're reading the file contents anyway to check whether modification is needed, so I have a slight preference for the existing code.

File: Library/Homebrew/cmd/configure-shell.rb
Comment (sjackman) [reply Library/Homebrew/cmd/configure-shell.rb:44]: I understand this argument, but I'm happy with what we have. The user will most likely run `brew configure-shell` only once at the time of installation.

**Review by sjackman — APPROVED** [review]

**Comment by sjackman** [comment]
The last run of CircleCI took 27 minutes to run. That's unusually long.
https://circleci.com/gh/Linuxbrew/brew/1218

**Comment by sjackman** [comment]
Thank you for this most excellent contribution to Linuxbrew, James! 🏆

**Comment by maxim-belkin** [comment]
> The last run of CircleCI took 27 minutes to run

It did a very thorough test, I guess.

**Pull request merged**
